### PR TITLE
Refactored datashader operations and add new aggregators

### PR DESCRIPTION
**Pull request by philippjfr**
Some refactoring of datashader operations to make it easier to implement new operations. Additionally adds an operation to aggregate ``Spikes``:

Handles simple spikes:

uniform_spikes_simple

Handles datetimes:

date_spikes_simple

Handles Spikes with vdims:

date_spikes_with_vdim

- [x] Add tests
- [x] Add Spikes aggregator with and with values and positive and negative heights
- [x] Add Area aggregator
- [x] Add Spread aggregator
- [x] Add docs

Cc: @jlstevens @jbednar

### CODE REVIEW

**Comment by philippjfr** [comment]
>so I'm not sure if it's worth implementing this more generally here already in anticipation?

I haven't changed anything about this and don't intend to.

**Comment by philippjfr** [comment]
Mostly wanted to do this so I can more easily implement the Segments aggregator in #3532.

**Comment by philippjfr** [comment]
Do need to know whether this addresses @jlstevens' needs though. I think to do what he was doing you need to create an NdOverlay of rasterized spikes with each with a fixed y_range and a RangeX stream.

**Comment by philippjfr** [comment]
Screen Shot 2019-08-18 at 11 23 03 PM

**Comment by philippjfr** [comment]
Would also like to compare performance to Jean-Luc's implementation, here's 10x1M spikes:

spikes

**Comment by jbednar** [comment]
> I haven't changed anything about this and don't intend to.

Ok. It showed up in the new stuff, but was probably just reformatted.  Your description of of Jean-Luc's needs sounds accurate to me, but I'll let him weigh in on that.

**Comment by philippjfr** [comment]
Updated the gif, dask appears slower than pure pandas here.

**Comment by philippjfr** [comment]
Once we have segments it'll probably might be more (computationally but not memory) efficient to precompute a Segments element to represent Jean-Luc's plot type, because it can represent the upper and lower end of each event.

**Comment by philippjfr** [comment]
>It showed up in the new stuff, but was probably just reformatted.

Yeah, lots of refactoring happening in this PR.

**Comment by jonmmease** [comment]
Am I right that this should make datashading the `Area` and `Spread` elements easier?  When I looked at it briefly a few months ago it looked like handling a variable number of kdims didn't fit well into the old structure.

**Comment by philippjfr** [comment]
>Am I right that this should make datashading the Area and Spread elements easier?

In some ways yes, but I'll probably have to do some more refactoring to get those done. That said I might tackle those in this PR as well.

File: holoviews/operation/datashader.py [hunk 1]
```diff
@@ -423,92 +546,69 @@ def _aggregate_ndoverlay(self, element, agg_fn):
...
-            return self.p.element_type(eldata, **params)
-        else:
-            layers = {}
-            for c in agg.coords[column].data:
-                cagg = agg.sel(**{column: c})
-                eldata = cagg if ds_version > '0.5.0' else (xs, ys, cagg.data)
-                layers[c] = self.p.element_type(eldata, **dict(params, vdims=vdims))
-            return NdOverlay(layers, kdims=[data.get_dimension(column)])
+        # Replacing x and y coordinates to avoid numerical precision issues
+        eldata = agg if ds_version > '0.5.0' else (xs, ys, agg.data)
```
Comment (philippjfr) [inline holoviews/operation/datashader.py]: Should remove this, this type of line aggregation wasn't supported in 0.5.0.

**Comment by jlstevens** [comment]
This is fantastic, thanks!

From the title of the PR, I was still expecting to implement spikes support on top of the refactored code so it is a pleasant surprise to see that `Spikes` support is what is being added here!

> I think to do what he was doing you need to create an NdOverlay of rasterized spikes with each with a fixed y_range and a RangeX stream.

That is exactly the approach I am currently using. As for performance, I'll check out this branch to compare but I doubt there will be much difference. If what I am currently using is significantly faster for some reason, I'll report back.

**Comment by philippjfr** [comment]
Area and Spread aggregation now also done, will add some tests and then document those in the Large Data guide.

**Comment by philippjfr** [comment]
Screen Shot 2019-08-20 at 12 24 16 AM

**Comment by philippjfr** [comment]
Ready to review and merge.

**Comment by jbednar** [comment]
Cool, thanks!

File: examples/user_guide/15-Large_Data.ipynb [hunk 2]
```diff
@@ -363,7 +363,13 @@
...
+    "As you can see in the examples above, converting the data to an image using Datashader makes it feasible to work with even very large datasets interactively.  One unfortunate side effect is that the original datapoints and line segments can no longer be used to support \"tooltips\" or \"hover\" information directly for RGB images generated with `datashade`; that data simply is not present at the browser level, and so the browser cannot unambiguously report information about any specific datapoint. \n",
+    "\n",
+    "Additionally once the ``shade`` operation is applied any detail about the underlying data is obscured. In order to get hover information there are therefore two options:\n",
```
Comment (jbednar) [inline examples/user_guide/15-Large_Data.ipynb]:
```suggestion
    "If you do need hover information, there are two good options available:\n",
```

File: examples/user_guide/15-Large_Data.ipynb [hunk 3]
```diff
@@ -363,7 +363,13 @@
...
+    "\n",
+    "1) Use the ``rasterize`` operation without `shade`\n",
```
Comment (jbednar) [inline examples/user_guide/15-Large_Data.ipynb]:
```suggestion
    "1) Use the ``rasterize`` operation without `shade`, which will let the plotting code handle the conversion to colors while still having the actual aggregated data to support hovering\n",
```

**Comment by philippjfr** [comment]
Will add some more tests in the morning.

**Comment by jlstevens** [comment]
@philippjfr Thanks for all this! I see you have now added some unit tests and the travis tests are green. Merging.

**Pull request merged**